execution_engine/src/core/runtime/host_function_flag.rs

```diff
@@ -0,0 +1,95 @@
+use std::{cell::Cell, rc::Rc};
+
+use tracing::error;
+
+/// A flag to indicate whether the current runtime call is made within the scope of a host function.
+///
+/// The flag is backed by an `Rc<Cell<u64>>`, meaning that clones will all share state.
+#[derive(Default, Clone)]
+pub(super) struct HostFunctionFlag {
+    /// A counter which, if non-zero, indicates that the `HostFunctionFlag` is `true`.
+    counter: Rc<Cell<u64>>,
+}
+
+impl HostFunctionFlag {
+    /// Returns `true` if this `HostFunctionFlag` has entered any number of host function scopes
+    /// without having exited them all.
+    pub(super) fn is_in_host_function_scope(&self) -> bool {
+        self.counter.get() != 0
+    }
+
+    /// Must be called when entering a host function scope.
+    ///
+    /// The returned `ScopedHostFunctionFlag` must be kept alive for the duration of the host
+    /// function call.  While at least one such `ScopedHostFunctionFlag` exists,
+    /// `is_in_host_function_scope()` returns `true`.
+    #[must_use]
+    pub(super) fn enter_host_function_scope(&self) -> ScopedHostFunctionFlag {
+        let new_count = self.counter.get().checked_add(1).unwrap_or_else(|| {
+            error!("checked_add failure in host function flag counter");
+            u64::MAX
+        });
+        self.counter.set(new_count);
+        ScopedHostFunctionFlag {
```

Comment:
We could probably remove `derive(Clone)` to avoid cloning outside of this module

Reply:
Well, I believe we still clone the `HostFunctionFlag` inside the `Runtime::new_from_self()` constructor.